pylint/checkers/modified_iterating_checker.py

```diff
@@ -72,7 +72,15 @@ def _modified_iterating_check(
         self, node: nodes.NodeNG, iter_obj: nodes.NodeNG
     ) -> None:
         msg_id = None
-        if self._modified_iterating_list_cond(node, iter_obj):
+        if self._deleted_iteration_target_cond(node, iter_obj):
```

Comment:
Not sure if this should be done in the same PR, but we're calling `safe_infer` quite a lot here. We probably hit the cache on every call but just to be safe:
shouldn't we just do `safe_infer` once before all the `ifs` and `elif` in between L75-87 and call methods accordingly?

Reply:
Good Q. I think the case where `safe_infer` is called multiple times is not tremendously realistic--just tuples and custom iterables that don't subclass list/dict/set, right? I should add a test case that relates to this PR, but the usefulness of a refactor is probably a separate PR.

Reply:
Never mind, I follow you now. Yeah that should be refactored in another PR. 👍🏻

Reply:
Ah, hang on. Just took a look -- it's different items being inferred. Only the list and the set branches (in the existing code) share a `safe_infer` call, and nothing is shared with the call in this PR. I'm not sure it's worth optimizing further -- we might be adding calls to `safe_infer` rather than minimizing them.